File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.cpp
#include <string>
#include <vector>

using namespace std;

long long MaxSum(vector<int> v){
    vector<long long> dp(v.size());
    dp[0] = v[0];
    
    long long Max = dp[0];
    for(int i = 1; i<dp.size();i++){
        if(dp[i - 1] + v[i] < v[i]) dp[i] = v[i];
        else dp[i] = dp[i-1] + v[i];
        
        Max = max(Max,(long long)dp[i]);
    }
    return Max;
    
    
}

long long solution(vector<int> sequence) {
    long long answer = 0;
    vector<int> v1(sequence), v2(sequence);
    
    for(int i = 0; i < sequence.size(); i++){
        if(i % 2 == 0) v1[i] *= -1;
        else v2[i] *= -1;
    }
    answer = max(MaxSum(v1), MaxSum(v2));
    return answer;
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    static long max;
    public long solution(int[] sequence) {
        long answer = 0;
        int n = sequence.length;
        max = Long.MIN_VALUE;
        
        int[] v1 = new int[n];
        int[] v2 = new int[n];
        long[] dp = new long[n];
        
        int x1 = 1;
        int x2 = -1;
        for(int i = 0; i < n; i++){
            v1[i] = sequence[i] * x1;
            v2[i] = sequence[i] * x2;
            x1 *= -1; x2 *=-1;
        }
        maxValue(v1, dp); maxValue(v2, dp);
        
        answer = max;
        
        return answer;
    }
    static void maxValue(int[] v, long[] dp){
        dp[0] = v[0];
        max = Math.max(max, v[0]);
        for(int i = 1; i < v.length; i++){
            dp[i] = Math.max(dp[i-1] + v[i], v[i]);
            max = Math.max(max,dp[i]);
        }
    }
        
}